### PR TITLE
I2C bus fixes for F1 and F4 targets

### PR DESCRIPTION
**Pull request by digitalentity**
9ff56d3 introduced very aggressive error handling, even a NAK (a normal bus event) was treated as a bus error and resulted in "unstick" procedute. Additionally, unstick was only clocking out 8 bits, while in most cases it should be 9 (8 bit data and ACK/NAK) and clock was followed by an empty trasnaction (START-STOP with no clock).

 Some of chips were stunned by these events and stopped responding to normal transactions - #665. Compass errors might be related as well - #632 (CC3D), #594 (NAZE)

### CODE REVIEW

**Comment by DzikuVx** [comment]
@digitalentity flashed my Flip32 and both Baro and Mag were detected. Good job

File: src/main/drivers/bus_i2c_stm32f10x.c [hunk 1]
```diff
@@ -132,11 +132,14 @@ void I2C3_EV_IRQHandler(void) {
 }
 #endif
 
-static bool i2cHandleHardwareFailure(I2CDevice device)
+static bool i2cHandleHardwareFailure(I2CDevice device, bool unstick)
```
Comment (martinbudden) [inline src/main/drivers/bus_i2c_stm32f10x.c]: I think it is better to pass in the `state` pointer - more flexible and also avoids a horrible `bool` parameter.

Comment (digitalentity) [reply src/main/drivers/bus_i2c_stm32f10x.c]: Agreed, fixing

File: src/main/drivers/bus_i2c.h [hunk 2]
```diff
@@ -51,6 +51,7 @@ typedef struct i2cDevice_s {
 typedef struct i2cState_s {
     volatile bool initialised;
     volatile bool error;
+    volatile bool unstick;
```
Comment (martinbudden) [inline src/main/drivers/bus_i2c.h]: `busError` is a better name than `unstick`

**Comment by martinbudden** [comment]
Great job fixing this. We've been getting far too many compass problems lately and hopefully this will help that.

**Comment by digitalentity** [comment]
Yes, I have a feeling that failed compass detection might be caused by the same aggressive error handling. It was mostly affecting the detection when NAKs are common for non-present devices.

**Comment by martinbudden** [comment]
@hydra you may wish to apply this fix. It appears PR https://github.com/cleanflight/cleanflight/pull/2310 causes some problems.

**Comment by hydra** [comment]
@martinbudden any chance of a PR for Cleanflight?

**Comment by martinbudden** [comment]
@hydra , OK, I'll look into it.

**Comment by digitalentity** [comment]
@martinbudden if you are not working on it - I have a minute to spare on porting this to CF.

**Comment by martinbudden** [comment]
@digitalentity , feel free to do this. I've had a brief look and it is not entirely straightforward, since Cleanflight does not use the new IO for I2C yet.

But it would also be good to get the 1.3 issues/PRs finished so we can do a release candidate for v1.3

**Comment by digitalentity** [comment]
Yes, that's a nuisance, but the fix is fairly simple, shouldn't be very hard to do. It's my fix after all :wink:

**Comment by digitalentity** [comment]
@hydra done

**Pull request merged**